### PR TITLE
Wsi to polygon

### PR DESCRIPTION
**Pull request by jonasteuwen**
- Allow convert_annotations function to create holes in polygon
- Create an util to convert mask to polygon, and CLI

### CODE REVIEW

**Comment by YoniSchirris** [comment]
@jonasteuwen All tests are failing

File: dlup/_region.py [hunk 1]
```diff
@@ -4,13 +4,13 @@
...
-from typing import Iterable, Union, cast
+from typing import Iterable, cast
 
 import numpy as np
 import PIL.Image
 
-_GenericFloatArray = Union[np.ndarray, Iterable[float]]
-_GenericIntArray = Union[np.ndarray, Iterable[int]]
+_GenericFloatArray = np.ndarray | Iterable[float]
+_GenericIntArray = np.ndarray | Iterable[int]
```
Comment (YoniSchirris) [inline dlup/_region.py:13]: Can we move these to `dlup.utils.typing` or something?

**Comment by YoniSchirris** [comment]
In the future can we keep PRs focused on the feature that is being implemented? 80% of the changes are typing refactors, massively distracting from the actual functional changes. Need to dig to find functional code changes.

File: dlup/utils/mask.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# Copyright (c) dlup contributors
+"""Utilities to work with binary masks"""
+import cv2
+import numpy as np
+import shapely.affinity
+from shapely.geometry import Polygon
+
+
+def _DFS(polygons, contours, hierarchy, sibling_id, is_outer, siblings, offset=(0, 0), scaling=1.0) -> list[Polygon]:
+    # Adapted FROM: https://gist.github.com/stefano-malacrino/7d429e5d12854b9e51b187170e812fa4
```
Comment (YoniSchirris) [inline dlup/utils/mask.py:11]: Missing docstring

Comment (YoniSchirris) [reply dlup/utils/mask.py:11]: Missing input typing

File: dlup/utils/mask.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    -------
+    list
+        The list of generated Shapely polygons
+    """
+    contours, hierarchy = cv2.findContours(mask, cv2.RETR_CCOMP,  cv2.CHAIN_APPROX_SIMPLE)
+
+    hierarchy = hierarchy[0]
+    polygons = []
+    _DFS(polygons, contours, hierarchy, 0, True, [], offset=offset, scaling=scaling)
+
```
Comment (YoniSchirris) [inline dlup/utils/mask.py:57]: can we do this in a more functional manner, such that it returns a new object instead of alters the existing `polygons` object which is an empty list?

This is difficult to read.

`_DFS` is also not a meaningful name at all; I have no clue what happens here

Comment (YoniSchirris) [reply dlup/utils/mask.py:57]: and add the parameter names in the function call so that from this call we know what `0` and `True` and `[]` represents.

Comment (YoniSchirris) [inline dlup/utils/mask.py:57]: Generally no clue what happens here from reading the code.

Can we add a test case to show it works?

File: dlup/cli/mask.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+from dlup.annotations import AnnotationType, Polygon, WsiAnnotations, WsiSingleLabelAnnotation
+from dlup.cli import file_path
+from dlup.data.dataset import TiledROIsSlideImageDataset
+from dlup.experimental_backends import ImageBackend
+from dlup.tiling import TilingMode
+from dlup.utils.mask import mask_to_polygons
+from functools import partial
+
+
+def _get_sample(index, dataset, index_map, scaling):
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:23]: - Missing typing
- Missing docstring

File: dlup/cli/mask.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    _mask = np.asarray(sample["image"])
+    for index in index_map:
+        curr_mask = (_mask == index).astype(np.uint8)
+        if curr_mask.sum() == 0:
+            continue
+        output[index_map[index]] = mask_to_polygons(_mask, offset=sample["coordinates"], scaling=scaling)
+    return output
+
+
+def dataset_to_polygon(dataset, index_map, num_workers=0, scaling=1.0, show_progress=True):
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:35]: - Missing return type
- Missing docstring

File: dlup/cli/mask.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                    pbar.update()
+                    for polygon_name in output_polygons:
+                        if polygon_name in curr_polygons:
+                            output_polygons[polygon_name] += curr_polygons[polygon_name]
+
+    geometry = {k: unary_union(polygons) for k, polygons in output_polygons.items()}
+    return geometry
+
+
+def mask_to_polygon(args: argparse.Namespace):
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:59]: - Missing return type

File: dlup/cli/mask.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    """Perform the mask conversion to polygon."""
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:60]: - Docstring very short for the seeming complexity of the function

File: dlup/cli/mask.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+
+def register_parser(parser: argparse._SubParsersAction):
+    """Register mask commands to a root parser."""
+    wsi_parser = parser.add_parser("mask", help="WSI mask parser")
+    wsi_subparsers = wsi_parser.add_subparsers(help="WSI mask subparser")
+    wsi_subparsers.required = True
+    wsi_subparsers.dest = "subcommand"
+
+    # Tile a slide and save the tiles in an output folder.
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:135]: This appears to be a comment from a copy-paste from the `wsi tiling` argparser.

Please delete

File: dlup/cli/mask.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    mask_parser = wsi_subparsers.add_parser(
+        "mask-to-polygon", help="Convert a WSI defining a mask to a GeoJSON polygon."
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:137]: is it sensical to term a mask `.tiff` a `WSI`?

File: dlup/cli/mask.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    )
+    mask_parser.add_argument(
+        "MASK_FN",
+        type=file_path,
+        help="Filename of the mask. If `--separate` is set, will create a label <MASK_FN>-<label>.json",
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:142]: can we write `filename` in full? or term it `path`?

There are several other locations in DLUP where we use `MASK_FN` with which we mean `MASK_FUNCTION`

File: dlup/cli/mask.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    )
+    mask_parser.add_argument(
+        "OUTPUT_FN",
+        type=lambda x: file_path(x, need_exists=False),
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:146]: shouldn't `type` just be `str` here?  or `file_path` (where ever that type is defined?) as on line 141.

File: dlup/cli/mask.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        type=int,
+        default=1024,
+        help="Size of the generated tiles.",
+    )
+    mask_parser.add_argument(
+        "--tile-overlap",
+        type=int,
+        default=128,
+        help="Number of overlapping pixels between tiles.",
+    )
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:175]: when/how do we need tile size and overlap? 

Do we generate tiles?

I thought we took a `mask.tiff` and transformed them to geojson object. I don't understand why we need to generate tiles.

File: dlup/data/transforms.py [hunk 13]
```diff
@@ -80,19 +80,29 @@ def convert_annotations(
...
             continue
 
+        original_values = None
+        interiors = [np.asarray(pi.coords).round().astype(np.int32) for pi in curr_annotation.interiors]
+        if interiors is not []:
+            original_values = mask.copy()
+            holes_mask = np.zeros(region_size, dtype=np.int32)
+            # Get a mask where the holes are
+            cv2.fillPoly(holes_mask, interiors, 1)
+
```
Comment (YoniSchirris) [inline dlup/data/transforms.py:90]: Can this be explained/described somewhere? Either in comments or docstring? I don't know what this adds to the function now.

**Review by YoniSchirris — CHANGES_REQUESTED** [review]
- Many changes are unrelated to the PR and contain typing changes, making it difficult to find functional changes that actually need review
- Many functional additions are missing typing, docstring, and inline comments for documentation
- checks are failing
- CLI appears to have arguments that are not required (e.g. tile size for writing tiles?)
- Test case is missing; Can we write a test that transforms a simple and complex binary (or multi labelled) mask to polygons to show in which cases it works? This can also function as `CLI` documentation.

File: dlup/cli/mask.py [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+from functools import partial
+
+
+def _get_sample(index, dataset, index_map, scaling):
+    output = {}
+    sample = dataset[index]
+    _mask = np.asarray(sample["image"])
+    for index in index_map:
+        curr_mask = (_mask == index).astype(np.uint8)
+        if curr_mask.sum() == 0:
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:29]: `not curr_mask.any()` may be faster

https://stackoverflow.com/questions/18395725/test-if-numpy-array-contains-only-zeros

File: dlup/cli/mask.py [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+            name, index = pair.split("=")
+            if not index.isnumeric():
+                raise argparse.ArgumentTypeError(f"Expected a key-pair of the form 1=tumor,2=stroma")
+            index = float(index)
+            if not index.is_integer():
+                raise argparse.ArgumentTypeError(f"Expected a key-pair of the form 1=tumor,2=stroma")
+            index = int(index)
+            if index == 0:
+                raise argparse.ArgumentTypeError(f"0 is not a proper index. Needs to be at least 1.")
+            index_map[index] = name.strip()
```
Comment (YoniSchirris) [inline dlup/cli/mask.py:100]: can we move this to a generic function in `utils.cli`? seems not related to the actual functionality of wsi-to-polygon, but just an argparse thing.

File: dlup/_image.py [hunk 16]
```diff
@@ -82,11 +92,20 @@ class SlideImage:
...
         self._identifier = identifier
 
+        if kwargs.get("interpolator", None) is not None:
+            interpolator = kwargs["interpolator"]
+            if isinstance(interpolator, Resampling):
+                interpolator = interpolator.name
+
+            self._interpolator = PIL.Image.Resampling[interpolator]
+        else:
+            self._interpolator = PIL.Image.Resampling.LANCZOS
```
[the remark this answers is not in the document]
Comment (jonasteuwen) [reply dlup/_image.py:107]: Yes, typically we don't add loggers in tested libraries.

File: dlup/_image.py [hunk 17]
```diff
@@ -82,11 +92,20 @@ class SlideImage:
     >>> wsi = dlup.SlideImage.from_file_path('path/to/slide.svs')
     """
 
-    def __init__(self, wsi: AbstractSlideBackend, identifier: Union[str, None] = None, **kwargs):
+    def __init__(self, wsi: AbstractSlideBackend, identifier: str | None = None, **kwargs):
```
[the remark this answers is not in the document]
Comment (jonasteuwen) [reply dlup/_image.py:95]: Right, but I'm in favor of consistency...

**Comment by rharkes** [comment]
To replace [Line 103-110 in mask.py](https://github.com/NKI-AI/dlup/blob/wsi-to-polygon/dlup/cli/mask.py#L103-L110)
```
    wsi_annotations = []
    for label in polygons:
        if isinstance(polygons[label], shapely.geometry.polygon.Polygon):
            wsi_annotations.append(
                WsiSingleLabelAnnotation(
                    label=label,
                    type=AnnotationType.POLYGON,
                    coordinates=[Polygon(polygons[label], label=label)],
                )
            )
        elif isinstance(polygons[label], shapely.geometry.multipolygon.MultiPolygon):
            wsi_single_label_annotation = WsiSingleLabelAnnotation(
                label=label,
                type=AnnotationType.POLYGON,
                coordinates=[Polygon(polygons[label].geoms[0], label=label)],
            )
            for i_polygons in range(1, len(polygons[label].geoms)):
                wsi_single_label_annotation.append(Polygon(polygons[label].geoms[i_polygons], label=label))
            wsi_annotations.append(wsi_single_label_annotation)
        else:
            raise NotImplementedError(
                f'Shapely type {type(polygons[label])} is not implemented. Region named "{label}" cannot be imported.'
            )
    slide_annotations = WsiAnnotations(wsi_annotations)
```

**Comment by rharkes** [comment]
[Line 31 in mask.py](https://github.com/NKI-AI/dlup/blob/wsi-to-polygon/dlup/cli/mask.py#L31)
`        output[index_map[index]] = mask_to_polygons(curr_mask, offset=sample["coordinates"], scaling=scaling)`

**Comment by rharkes** [comment]
at the end of mask.py
```
    mask_parser.add_argument(
        "--simplify",
        type=float,
        default=-1.0,
        help="The maximum allowed geometry displacement. The higher this value, the smaller the number of vertices in the resulting geometry. Set to -1 to not simplify the shapes. Default: -1",
    )
```
and to apply simplify:
```
    slide_annotations = WsiAnnotations(wsi_annotations)
    if not args.simplify == -1:
        slide_annotations.simplify(tolerance=args.simplify)
```

File: dlup/cli/mask.py (hunk 11 above)
Comment (jonasteuwen) [reply dlup/cli/mask.py:119]: No, then it will automatically cast as path

File: dlup/cli/mask.py (hunk 10 above)
Comment (jonasteuwen) [reply dlup/cli/mask.py:115]: Yes

File: dlup/cli/mask.py (hunk 9 above)
Comment (jonasteuwen) [reply dlup/cli/mask.py:110]: Not really no, fixed

File: dlup/utils/mask.py (hunk 2 above)
Comment (jonasteuwen) [reply dlup/utils/mask.py]: Did some. It's not trivial here, later on we will do strict mypy

File: dlup/_region.py (hunk 1 above)
Comment (jonasteuwen) [reply dlup/_region.py]: Yeah, sometimes it's better to repeat
